metadata/src/main/java/org/apache/kafka/controller/MetaLogManager.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ApiMessageAndVersion;
+
+/**
+ * The MetaLogManager handles storing metadata and electing leaders.
+ */
+public interface MetaLogManager extends AutoCloseable {
+    /**
+     * Listeners receive notifications from the MetaLogManager.
+     */
+    interface Listener {
+        /**
+         * Called when the MetaLogManager commits a message.
+         *
+         * @param epoch         The controller epoch of the message.
+         * @param index         The index of the message within the controller epoch.
+         * @param message       The message.
+         */
+        void commit(long epoch, long index, ApiMessage message);
+
+        /**
+         * Called when the MetaLogManager has claimed the leadership.
+         *
+         * @param epoch         The controller epoch that is starting.
+         */
+        void claim(long epoch);
+
+        /**
+         * Called when the MetaLogManager has renounced the leadership.
+         *
+         * @param epoch         The controller epoch that has ended.
+         */
+        void renounce(long epoch);
+
+        /**
+         * Called when the MetaLogManager has finished shutting down, and wants to tell its
+         * listener that it is safe to shut down as well.
+         */
+        void beginShutdown();
+    }
+
+    /**
+     * Schedule a write to the log.
+     *
+     * The write will be scheduled to happen at some time in the future.  There is no
+     * error return or exception thrown if the write fails.  Instead, the listener may
+     * regard the write as successful if and only if the MetaLogManager reaches the given
+     * index before renouncing its leadership.
```

Comment:
`if the MetaLogManager reaches the given index` was confusing to me, perhaps because there is no index given here.  It took a moment for me to realize that what you are saying (I think?) is that a listener may regard the write as successful if and only the MetaLogManager reaches the returned index and notifies the listener via `commit()`.  Is that right, and might this be worded more clearly?

Reply:
I added a sentence about monitoring the committed offsets.  I'm open to how we could rephrase this.  Not sure how to explain it better.